db/channel_cache.go

```diff
@@ -214,831 +245,202 @@ func (c *channelCacheImpl) ReleaseLateSequenceClient(channelName string, current
+//	//     1. addChannelCache issued for channel A
+//	//     2. addChannelCache obtains stable sequence, seq=10
+//	//     3. addToCache (from another goroutine) receives sequence 11 in channel A, but detects that it's inactive (not in c.channelCaches)
+//	//     4. addChannelCache initializes cache with validFrom=10 and adds to c.channelCaches
+//	//  This scenario would result in sequence 11 missing from the cache.  Locking seqLock ensures that
+//	//  step 3 blocks until step 4 is complete (and so sees the channel as active)
+func (c *channelCacheImpl) addChannelCache(channelName string) *singleChannelCache {
 
-func newChannelCacheWithOptions(context *DatabaseContext, channelName string, validFrom uint64, options ChannelCacheOptions) *singleChannelCache {
-	cache := newChannelCache(context, channelName, validFrom)
+	// TODO: Lock seqLock instead of cacheLock
+	c.seqLock.Lock()
 
-	// Update cache options when present
-	if options.ChannelCacheMinLength > 0 {
-		cache.options.ChannelCacheMinLength = options.ChannelCacheMinLength
-	}
+	// Everything after the current high sequence will be added to the cache via the feed
+	validFrom := c._getHighCacheSequence() + 1
 
-	if options.ChannelCacheMaxLength > 0 {
-		cache.options.ChannelCacheMaxLength = options.ChannelCacheMaxLength
-	}
+	singleChannelCache := newChannelCacheWithOptions(c.queryHandler, channelName, validFrom, c.options, c.statsMap)
+	cacheValue, created, cacheSize := c.channelCaches.GetOrInsert(channelName, singleChannelCache)
+	c.seqLock.Unlock()
 
-	if options.ChannelCacheAge > 0 {
-		cache.options.ChannelCacheAge = options.ChannelCacheAge
-	}
+	singleChannelCache = AsSingleChannelCache(cacheValue)
 
-	if options.ChannelCacheMaxNumber > 0 {
-		cache.options.ChannelCacheMaxNumber = options.ChannelCacheMaxNumber
+	if !c.isCompactActive() && cacheSize > c.compactHighWatermark {
+		c.startCacheCompaction()
 	}
 
-	base.Infof(base.KeyCache, "Initialized cache for channel %q with options: %+v, validFrom: %d", base.UD(cache.channelName), cache.options, validFrom)
-
-	return cache
-}
-
-type ChannelCacheOptions struct {
-	ChannelCacheMinLength int           // Keep at least this many entries in cache
-	ChannelCacheMaxLength int           // Don't put more than this many entries in cache
-	ChannelCacheAge       time.Duration // Keep entries at least this long
-	ChannelCacheMaxNumber int           // Maximum number of channel caches which will exist at any one point
-}
-
-// Low-level method to add a LogEntry to a single channel's cache.
-func (c *singleChannelCache) addToCache(change *LogEntry, isRemoval bool) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-
-	if c.wouldBeImmediatelyPruned(change) {
-		base.Infof(base.KeyCache, "Not adding change #%d doc %q / %q ==> channel %q, since it will be immediately pruned",
-			change.Sequence, base.UD(change.DocID), change.RevID, base.UD(c.channelName))
-		return
+	if created {
+		c.statsMap.Add(base.StatKeyChannelCacheNumChannels, 1)
```

Comment:
Do we need to decrement this stat on eviction? Or are we treating it as a "Total num of channel caches created ever" type stat?

Reply:
Good catch - will add this.